### PR TITLE
feat(amazonq): Add command to generate unit tests for selected code

### PR DESCRIPTION
**Pull request by bweedop**
Adds command to Amazon Q to generate unit tests for selected code. The feature will be limited to internal Amazon users to collect usage metrics and, subsequently, will be released to all users.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

## Description

- Adds "Test" command to context menu (with a (Beta) suffix for while internal only). The command will send the selected code to the chat with the following prompt: "Generate unit tests for the following part of my code for me:"

image

## Checklist
- [X] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/codewhisperer/util/authUtil.ts
Comment (jpinkney-aws) [inline packages/core/src/codewhisperer/util/authUtil.ts:154]: is this something that should live in the general activation since it's not necessarily relevant to codewhisperer?

File: packages/core/package.nls.json
Comment (jpinkney-aws) [inline packages/core/package.nls.json:114]: Should this be something like `Generate Tests (Beta)` ? When I right click and see "Test" for some reason I think its going to automagically run my tests for me

File: packages/core/src/codewhispererChat/controllers/chat/userIntent/userIntentRecognizer.ts
Comment (hayemaxi) [inline packages/core/src/codewhispererChat/controllers/chat/userIntent/userIntentRecognizer.ts:44]: use `AuthUtil. isInternalAmazonUser()`

File: packages/core/src/codewhispererChat/commands/registerCommands.ts
Comment (laileni-aws) [inline packages/core/src/codewhispererChat/commands/registerCommands.ts:113]: nit: Can this be named more specific to unit test instead of generalized `testCode`?

File: packages/core/src/codewhisperer/util/authUtil.ts
Comment (laileni-aws) [inline packages/core/src/codewhisperer/util/authUtil.ts:201]: nitL Was this ok to expose startUrl to public?

Comment (bweedop) [reply packages/core/src/codewhisperer/util/authUtil.ts:201]: Yes. There has been a lot of back and forth on this (on Slack) and, ultimately, the decision is that this is okay.
 https://github.com/aws/aws-toolkit-jetbrains/pull/4869#discussion_r1750938610

File: packages/core/src/codewhispererChat/commands/registerCommands.ts
Comment (bweedop) [reply packages/core/src/codewhispererChat/commands/registerCommands.ts:113]: Yes

File: packages/core/package.nls.json
Comment (bweedop) [reply packages/core/package.nls.json:114]: Yes. I've confirmed with product that "Generate tests (Beta)" makes sense.

File: packages/core/src/codewhisperer/util/authUtil.ts
Comment (bweedop) [reply packages/core/src/codewhisperer/util/authUtil.ts:154]: I think this makes sense.

File: packages/core/src/codewhisperer/util/authUtil.ts
Comment (ivikash) [reply packages/core/src/codewhisperer/util/authUtil.ts:201]: Can we move this utility function to a common place that can be use by Amazon Q and Toolkits.

File: packages/core/src/shared/vscode/setContext.ts
Comment (justinmk3) [inline packages/core/src/shared/vscode/setContext.ts:17]: I think the key should be named after the feature. The fact that it's currently gated on "internal user" is not something that should be in the key name.

File: packages/core/src/auth/auth.ts
Comment (justinmk3) [inline packages/core/src/auth/auth.ts:248]: I was wondering if `forgetConnection` and `expireConnection` should also un-set this. But it's probably useful to keep this hint around, once it was deduced. And in that case, should this flag be stored in globalState? Then we don't need to wait until a valid connection next time.

File: packages/core/src/auth/auth.ts
Comment (justinmk3) [inline packages/core/src/auth/auth.ts:173]: would this comment make sense as the docstring for `reformatStartUrl` instead

File: packages/core/src/auth/auth.ts
Comment (justinmk3) [inline packages/core/src/auth/auth.ts:183]:
```suggestion
    public normalizeStartUrl(startUrl: string | undefined) {
```

**Review by justinmk3 — APPROVED** [review]
LGTM after comments. Needs approval from someone on your team too.

File: packages/core/src/auth/auth.ts
Comment (bweedop) [reply packages/core/src/auth/auth.ts:248]: I don't think it hurts to set this back to false on `forgetConnection` and `expireConnection`.

Some testing that I did made me doubt how `globalState` worked. Using `globalState.update` for this didn't work to enable the feature for internal users. `setContext`, however, worked. Are global keys detectable from package.json after being set?

**Review by laileni-aws — APPROVED** [review]

File: packages/core/src/auth/auth.ts
Comment (justinmk3) [reply packages/core/src/auth/auth.ts:248]: > Are global keys detectable from package.json after being set?

No

File: packages/core/src/auth/auth.ts
Comment (justinmk3) [reply packages/core/src/auth/auth.ts:173]: move the comment to normalizeStartUrl

File: packages/core/src/auth/auth.ts
Comment (bweedop) [reply packages/core/src/auth/auth.ts:248]: Oh okay well that would explain it. We need this to be set with `setContext`

File: packages/core/src/auth/auth.ts
Comment (bweedop) [reply packages/core/src/auth/auth.ts:173]: Done

**Review by dogusata — APPROVED** [review]

**Pull request merged**
